rs/registry/node_provider_rewards/src/v1_rewards.rs

```diff
@@ -0,0 +1,535 @@
+use ic_base_types::PrincipalId;
+use ic_protobuf::registry::node_rewards::v2::{NodeRewardRate, NodeRewardsTable};
+use itertools::Itertools;
+use num_traits::ToPrimitive;
+
+use crate::v1_logs::LogLevel;
+use crate::v1_types::TimestampNanos;
+use crate::{
+    v1_logs::{LogEntry, Operation, RewardsLog},
+    v1_types::{
+        DailyNodeMetrics, RegionNodeTypeCategory, RewardableNode, Rewards, RewardsPerNodeProvider,
+    },
+};
+use ic_management_canister_types::{NodeMetrics, NodeMetricsHistoryResponse};
+use rust_decimal::Decimal;
+use rust_decimal_macros::dec;
+use std::collections::HashMap;
+
+const FULL_REWARDS_MACHINES_LIMIT: u32 = 3;
+const MIN_FAILURE_RATE: Decimal = dec!(0.1);
+const MAX_FAILURE_RATE: Decimal = dec!(0.6);
+const MAX_REWARDS_REDUCTION: Decimal = dec!(0.8);
+const RF: &str = "Linear Reduction factor";
+
+pub fn calculate_rewards(
+    days_in_period: u64,
+    rewards_table: &NodeRewardsTable,
+    subnet_metrics: HashMap<PrincipalId, Vec<NodeMetricsHistoryResponse>>,
+    rewardable_nodes: &[RewardableNode],
+) -> RewardsPerNodeProvider {
+    let mut rewards_per_node_provider = HashMap::default();
+    let mut rewards_log_per_node_provider = HashMap::default();
+
+    let mut all_assigned_metrics = daily_node_metrics(subnet_metrics);
+    let subnets_systematic_fr = systematic_fr_per_subnet(&all_assigned_metrics);
+    let node_provider_rewardables = rewardables_by_node_provider(rewardable_nodes);
+
+    for (node_provider_id, node_provider_rewardables) in node_provider_rewardables {
+        let mut logger = RewardsLog::default();
+        logger.add_entry(
+            LogLevel::High,
+            LogEntry::CalculateRewardsForNodeProvider(node_provider_id),
+        );
+
+        let assigned_metrics: HashMap<PrincipalId, Vec<DailyNodeMetrics>> =
```

Comment:
I think it would help if you were using SubnetIds where possible.  It's hard to keep track of all the different meanings of keys in each map when they're all principals.  In some cases it's nodes, in others it's subnets, and in some it's node providers.